### PR TITLE
MediaService: Cleanup part 2, cacheing, URLs, and tests.

### PR DESCRIPTION
**Pull request by kurzee**
This is technically another cleanup PR as it clears the path by changing the URL and file handling of actual Media files on disk via the local Media directory.

For the most part the "Media" directory is used as cache, but it's also kind of used as a scratch disk for uploads. While this Media directory should actually be within the application "Caches" directory, we'll keep it within "Documents" for now as I refresh the actual handling of exported assets on and off disk, in the next PRs.

Notable changes:

1. `clearUnusedFilesFromLocalMediaDirectory` and `clearCachedFilesFromLocalMediaDirectory ` were formerly seen independently as `cleanUnusedMediaFilesFromMediaCacheFolder` and `cleanMediaCacheFolder `: https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/Services/MediaService.h#L175 
   - Which now share `purgeLocalMediaFiles` as the actual implementation.
2. New tests under `Services/MediaServiceTests.swift` and `Models/MediaTests.swift`.
3. `AppSettingsViewController.swift` now handles the newly async nature of both calculating and clearing the local media directory, via new methods.
   - Also moves the `MediaSizingRow` out of `WPImmuTableRows.swift`
4. Refactors `Media` absolute URL handling to use `NSURL` and be relative to the `MediaService.localMediaDirectory()`.

To test:

1. Run all project tests.

To test:

1. Test that inserting and posting an image from the device library works.
2. Test that inserting and posting an video from the device library works.
3. Test that inserting and posting an image from the Media library works.
4. Test that inserting and posting a video from the Media library works.
5. Test that inserting and posting an image that is pasted into the editor works.
6. Test that inserting and posting a GIF from the device library works.
7. Test that inserting and posting a GIF from the Media library works.

To test:
1. Go to Me > App Settings.
2. Clear the cache.
3. Check that the UI was async, or at least it didn't noticeably lock.
4. Most of the cache should be cleared, some items may remain.
5. Background app.
6. Kill app.
7. Relaunch, but look for the log `"Media: removed \(removedCount) file(s) during cleanup."` in the console.
8. If no log appeared, switch blogs and repeat step 5-7, the log should show up when switching blogs or every so often.

Needs review: @frosty can I get your eyes on this one? Also, cc @SergioEstevao to keep an eye on the deeper implementation changes.

Note: Media in general is still a WIP, with plenty still yet to change. If something looks a bit funky with naming, or separation of concerns, bring it up and I'll keep it in mind for the next PRs.

### CODE REVIEW

**Comment by frosty** [comment]
I'm getting a crash whenever I access my media library (either through the picker or the Media section):

```
Enqueued from com.apple.main-thread (Thread 1)Queue : com.apple.main-thread (serial)
#0	0x0000000116583188 in _dispatch_queue_push ()
#1	0x000000010c3fe383 in -[MediaService imageForMedia:size:success:failure:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/Services/MediaService.m:587
#2	0x000000010c2eb1ac in -[Media(WPMediaAsset) imageWithSize:completionHandler:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m:550
#3	0x0000000111a1925d in -[WPMediaCollectionViewCell setAsset:] at /Users/frosty/src/WordPress-iOS/Pods/WPMediaPicker/Pod/Classes/WPMediaCollectionViewCell.m:90
#4	0x0000000111a22ee4 in -[WPMediaPickerViewController collectionView:cellForItemAtIndexPath:] at /Users/frosty/src/WordPress-iOS/Pods/WPMediaPicker/Pod/Classes/WPMediaPickerViewController.m:341
#5	0x0000000110ae49b1 in -[UICollectionView _createPreparedCellForItemAtIndexPath:withLayoutAttributes:applyAttributes:isFocused:notify:] ()
#6	0x0000000110ae47ed in -[UICollectionView _createPreparedCellForItemAtIndexPath:withLayoutAttributes:applyAttributes:] ()
#7	0x0000000110b03220 in __51-[UICollectionView _viewAnimationsForCurrentUpdate]_block_invoke.1941 ()
#8	0x0000000110b00647 in -[UICollectionView _viewAnimationsForCurrentUpdate] ()
#9	0x0000000110b05c7f in __71-[UICollectionView _updateWithItems:tentativelyForReordering:animator:]_block_invoke.2012 ()
#10	0x000000011026ad3e in +[UIView(Animation) performWithoutAnimation:] ()
#11	0x0000000110b048b9 in -[UICollectionView _updateWithItems:tentativelyForReordering:animator:] ()
#12	0x0000000110afebbf in -[UICollectionView _endItemAnimationsWithInvalidationContext:tentativelyForReordering:animator:] ()
#13	0x0000000110b07259 in -[UICollectionView _endUpdatesWithInvalidationContext:tentativelyForReordering:animator:] ()
#14	0x0000000110b075a0 in -[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:animator:] ()
#15	0x0000000110b073c8 in -[UICollectionView _performBatchUpdates:completion:invalidationContext:tentativelyForReordering:] ()
#16	0x0000000110b0734a in -[UICollectionView _performBatchUpdates:completion:invalidationContext:] ()
#17	0x0000000110b0729f in -[UICollectionView performBatchUpdates:completion:] ()
#18	0x0000000111a1fbea in -[WPMediaPickerViewController updateDataWithRemoved:inserted:changed:moved:] at /Users/frosty/src/WordPress-iOS/Pods/WPMediaPicker/Pod/Classes/WPMediaPickerViewController.m:166
#19	0x0000000111a1ef01 in __42-[WPMediaPickerViewController viewDidLoad]_block_invoke at /Users/frosty/src/WordPress-iOS/Pods/WPMediaPicker/Pod/Classes/WPMediaPickerViewController.m:92
#20	0x000000010c2a3260 in __65-[WPAndDeviceMediaLibraryDataSource registerChangeObserverBlock:]_block_invoke.42 at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Media/WPAndDeviceMediaLibraryDataSource.m:107
#21	0x000000010c2e52f9 in -[MediaLibraryPickerDataSource notifyObserversWithIncrementalChanges:removed:inserted:changed:moved:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m:133
#22	0x000000010c2e916b in -[MediaLibraryPickerDataSource controllerDidChangeContent:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m:409
#23	0x000000011354a8bf in __82-[NSFetchedResultsController(PrivateMethods) _core_managedObjectContextDidChange:]_block_invoke ()
#24	0x0000000113425b52 in developerSubmittedBlockToNSManagedObjectContextPerform ()
#25	0x0000000113425a1f in -[NSManagedObjectContext performBlockAndWait:] ()
#26	0x0000000113548f07 in -[NSFetchedResultsController(PrivateMethods) _core_managedObjectContextDidChange:] ()
#27	0x00000001152c3c2c in __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ ()
#28	0x00000001152c3b29 in _CFXRegistrationPost ()
#29	0x00000001152c3892 in ___CFXNotificationPost_block_invoke ()
#30	0x0000000115287162 in -[_CFXNotificationRegistrar find:object:observer:enumerator:] ()
#31	0x00000001152862c1 in _CFXNotificationPost ()
#32	0x000000010f1e3b24 in -[NSNotificationCenter postNotificationName:object:userInfo:] ()
#33	0x000000011340ec65 in -[NSManagedObjectContext(_NSInternalNotificationHandling) _postObjectsDidChangeNotificationWithUserInfo:] ()
#34	0x00000001134a27ba in -[NSManagedObjectContext(_NSInternalChangeProcessing) _createAndPostChangeNotification:deletions:updates:refreshes:deferrals:wasMerge:] ()
#35	0x0000000113408e10 in -[NSManagedObjectContext(_NSInternalChangeProcessing) _processRecentChanges:] ()
#36	0x000000011342b8e2 in -[NSManagedObjectContext(_NestedContextSupport) _parentProcessSaveRequest:inContext:error:] ()
#37	0x00000001134a6927 in __82-[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:]_block_invoke ()
#38	0x00000001134260ea in internalBlockToNSManagedObjectContextPerform ()
#39	0x00000001165a705c in _dispatch_client_callout ()
#40	0x000000011658a0f8 in _dispatch_barrier_sync_f_slow_invoke ()
#41	0x00000001165a705c in _dispatch_client_callout ()
#42	0x000000011658840b in _dispatch_main_queue_callback_4CF ()
#43	0x00000001152e9909 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#44	0x00000001152afae4 in __CFRunLoopRun ()
#45	0x00000001152af016 in CFRunLoopRunSpecific ()
#46	0x00000001180c5a24 in GSEventRunModal ()
#47	0x00000001101b30d4 in UIApplicationMain ()
#48	0x000000010c1c11fe in main at /Users/frosty/src/WordPress-iOS/WordPress/main.m:5
#49	0x00000001165f365d in start ()
```

**Comment by kurzee** [comment]
>One issue I've noticed is that there's a retain cycle somewhere in `AppSettingsViewController`

@frosty found it! There were a few sketchy things in there. Also cleaned up the other code issues mentioned. Ready for another look at your leisure.

File: WordPress/Classes/Extensions/MediaService+Local.swift [hunk 1]
```diff
@@ -93,21 +93,21 @@ extension MediaService {
...
 
-    /// Clear the locala Media directory of any files that are no longer in use by any managed Media objects.
+    /// Clear the local Media directory of any files that are no longer in use by any managed Media objects.
     ///
     /// - Note: These files can show up because of the app being killed while a media object
     ///   was being created or when a CoreData migration fails and the database is recreated.
     ///
     class func clearUnusedFilesFromLocalMediaDirectory(onCompletion: (() -> ())?, onError: ((Error) -> Void)?) {
-        purgeLocalMediaFiles(exceptMedia: NSPredicate.init(format: "blog != NULL", argumentArray: nil),
+        purgeLocalMediaFiles(exceptMedia: NSPredicate(format: "blog != NULL", argumentArray: nil),
```
Comment (frosty) [inline WordPress/Classes/Extensions/MediaService+Local.swift]: Nitpick, but you can remove the `argumentArray` parameter entirely from these calls.

**Comment by kurzee** [comment]
Thanks @frosty!

**Comment by kurzee** [comment]
@SergioEstevao if you find anything from this PR that needs some extra care let me know and I'll bring it into one of the next PRs.

**Pull request merged**